### PR TITLE
Add FQN support for DB plugins

### PR DESCRIPTION
**Pull request by snehalakshmisha**
Jira: https://cdap.atlassian.net/browse/CDAP-19665

Testing:
FQNs generated were
1. Sql Server:
source - sqlserver://ip:1433/AdventureWorks2014.sqlServerSource3
sink - sqlserver://ip:1433/AdventureWorks2014.sqlServerSink3
2. MySql
source - mysql://ip:3306/sakila.mySqlSource0
sink - mysql://ip:3306/sakila.mySqlServerSink0
3. Postgres
source - postgresql://ip:5432/cdf.pgSource0
sink - postgresql://ip:5432/cdf.pgSink0
4. Oracle
source - oracle://ip:1521/ORCL.orclSource0
oracle://ip:1521/ORCL.orclSink0
location is 'global' by default for the above 4 on-prem instances
5. Cloudsql MySql
source -mysql://instName:3306/db.cloudsqlSource
location - us-east1(or one such region)
sink -mysql://instName:3306/db.cloudsqlSink
location - us-east1(or one such region)
6. CloudSql Postgres
source -postgresql://instName: 5432/db.cloudPgSource0
location - us-east1
sink -postgresql://instName: 5432/db.cloudPgSink0
location - us-east1

### CODE REVIEW

File: cloudsql-mysql-plugin/widgets/CloudSQLMySQL-batchsink.json [hunk 1]
```diff
@@ -98,7 +98,7 @@
           "label": "Reference Name",
           "name": "referenceName",
           "widget-attributes": {
-            "placeholder": "Name used to identify this sink for lineage"
+            "placeholder": "Name used to identify this source for lineage. Typically, the name of the table/view."
```
Comment (itsankit-google) [inline cloudsql-mysql-plugin/widgets/CloudSQLMySQL-batchsink.json]: sink?

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java [hunk 2]
```diff
@@ -71,6 +76,21 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
   protected SchemaReader getSchemaReader() {
     return new CommonSchemaReader();
   }
+
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSinkContext context) {
+    // connection is of the form <projectId>:<region>:<instanceName>
```
Comment (itsankit-google) [inline cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java]: for private cloudsql mysql instance it is different. 
ref - [CloudSQLMySQLConnectorConfig.java#L37](https://github.com/data-integrations/database-plugins/blob/develop/cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLConnectorConfig.java#L37)

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSource.java [hunk 3]
```diff
@@ -90,6 +94,21 @@ protected String createConnectionString() {
         cloudsqlMysqlSourceConfig.connection.getConnectionName());
   }
 
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSourceContext context) {
+    // connection is of the form <projectId>:<region>:<instanceName>
```
Comment (itsankit-google) [inline cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSource.java]: similar comment here.

File: cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java [hunk 4]
```diff
@@ -107,6 +111,21 @@ protected FieldsValidator getFieldsValidator() {
     return new PostgresFieldsValidator();
   }
 
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSinkContext context) {
+    // connection is of the form <projectId>:<region>:<instanceName>
```
Comment (itsankit-google) [inline cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java]: similar comment here.

File: cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSource.java [hunk 5]
```diff
@@ -97,6 +101,21 @@ protected String createConnectionString() {
         cloudsqlPostgresqlSourceConfig.connection.getConnectionName());
   }
 
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSourceContext context) {
+    // connection is of the form <projectId>:<region>:<instanceName>
```
Comment (itsankit-google) [inline cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSource.java]: similar comment here.

File: cloudsql-postgresql-plugin/widgets/CloudSQLPostgreSQL-batchsink.json [hunk 6]
```diff
@@ -98,7 +98,7 @@
           "label": "Reference Name",
           "name": "referenceName",
           "widget-attributes": {
-            "placeholder": "Name used to identify this sink for lineage"
+            "placeholder": "Name used to identify this source for lineage. Typically, the name of the table/view."
```
Comment (itsankit-google) [inline cloudsql-postgresql-plugin/widgets/CloudSQLPostgreSQL-batchsink.json]: sink?

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java [hunk 7]
```diff
@@ -71,6 +76,21 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
+    String[] connectionParams = cloudsqlMysqlSinkConfig.getConnection().getConnectionName().split(":");
+    String fqn = DBUtils.constructFQN(cloudsqlMysqlSinkConfig.getJdbcPluginName(),
```
Comment (itsankit-google) [inline cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java]: not sure if `jdbcPluginName` is a reliable way to identify `dbType` because it is configurable from user, I can change it to literally anything like `abc....`. 
ref - 
image

For specific db plugins, we can hardcode the input param.

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresSink.java [hunk 8]
```diff
@@ -94,6 +99,20 @@ protected FieldsValidator getFieldsValidator() {
...
 
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSinkContext context) {
+    // dbtype, host, port, db from the connection string
+    // table is the reference name
+    String fqn = DBUtils.constructFQN(postgresSinkConfig.getJdbcPluginName(),
+                                      postgresSinkConfig.getConnection().getHost(),
+                                      postgresSinkConfig.getConnection().getPort(),
+                                      postgresSinkConfig.getConnection().getDatabase(),
+                                      context.getInputSchema().getDisplayName(),
```
Comment (itsankit-google) [inline postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresSink.java]: This is not the correct schame name. It is the name of CDAP schema, we need to pass Database Schema here.
In sinks, we do have a config named as `databaseSchema` but that can be nullable.
In source, we might need to parse the import query.

Comment (snehalakshmisha) [reply postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresSink.java]: removed this for now since it requires parsing

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java [hunk 9]
```diff
@@ -71,6 +76,29 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
+    if (CloudSQLUtil.PRIVATE_INSTANCE.equalsIgnoreCase(cloudsqlMysqlSinkConfig.getConnection().getInstanceType())) {
+      // connection is the private IP address
+      ip = cloudsqlMysqlSinkConfig.getConnection().getConnectionName();
+    } else {
+      // connection is of the form <projectId>:<region>:<instanceName>
+      String[] connectionParams = cloudsqlMysqlSinkConfig.getConnection().getConnectionName().split(":");
+      ip = connectionParams[2];
+      location = connectionParams[1];
+    }
+    String fqn = DBUtils.constructFQN("mysql", ip, 0,
```
Comment (itsankit-google) [inline cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java]: Why `0`? it uses `3306` by default.

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java [hunk 10]
```diff
@@ -71,6 +76,29 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
   protected SchemaReader getSchemaReader() {
     return new CommonSchemaReader();
   }
+
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSinkContext context) {
+    String ip;
```
Comment (itsankit-google) [inline cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java]: let's call it `host` instead of `ip`.

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSource.java [hunk 11]
```diff
@@ -90,6 +95,29 @@ protected String createConnectionString() {
         cloudsqlMysqlSourceConfig.connection.getConnectionName());
   }
 
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSourceContext context) {
+    String ip;
```
Comment (itsankit-google) [inline cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSource.java]: similar comment here

File: database-commons/src/main/java/io/cdap/plugin/util/DBUtils.java [hunk 12]
```diff
@@ -280,6 +280,15 @@ private static void unregisterOracleMBean(ClassLoader classLoader) {
     }
   }
 
+  public static String constructFQN(String dbType, String host, int port, String db, String tableName) {
+    if (dbType.startsWith("cloud")) {
```
Comment (itsankit-google) [inline database-commons/src/main/java/io/cdap/plugin/util/DBUtils.java]: This `if` will always return `false` now right?

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSink.java [hunk 13]
```diff
@@ -74,6 +79,18 @@ protected FieldsValidator getFieldsValidator() {
     return new SqlFieldsValidator();
   }
 
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSinkContext context) {
+    // dbtype, host, port, db from the connection string
+    // table is the reference name
+    String fqn = DBUtils.constructFQN("sqlserver",
```
Comment (itsankit-google) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSink.java]: nit - `sqlserver` -> `mssql`

File: mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSource.java [hunk 14]
```diff
@@ -71,6 +75,18 @@ protected Class<? extends DBWritable> getDBRecordType() {
     return SqlServerSourceDBRecord.class;
   }
 
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSourceContext context) {
+    // dbtype, host, port, db from the connection string
+    // table is the reference name
+    String fqn = DBUtils.constructFQN("sqlserver",
```
Comment (itsankit-google) [inline mssql-plugin/src/main/java/io/cdap/plugin/mssql/SqlServerSource.java]: similarly here.

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSource.java [hunk 15]
```diff
@@ -90,6 +95,29 @@ protected String createConnectionString() {
...
+    if (CloudSQLUtil.PRIVATE_INSTANCE.equalsIgnoreCase(cloudsqlMysqlSourceConfig.getConnection().getInstanceType())) {
+      // connection is the private IP address
+      ip = cloudsqlMysqlSourceConfig.getConnection().getConnectionName();
+    } else {
+      // connection is of the form <projectId>:<region>:<instanceName>
+      String[] connectionParams = cloudsqlMysqlSourceConfig.getConnection().getConnectionName().split(":");
+      ip = connectionParams[2];
+      location = connectionParams[1];
+    }
+    String fqn = DBUtils.constructFQN("mysql", ip, 0,
```
Comment (itsankit-google) [inline cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSource.java]: similarly use `3306`

File: cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java [hunk 16]
```diff
@@ -107,6 +112,29 @@ protected FieldsValidator getFieldsValidator() {
...
+    if (CloudSQLUtil.PRIVATE_INSTANCE.equalsIgnoreCase(cloudsqlPostgresqlSinkConfig.getConnection().getInstanceType())) {
+      // connection is the private IP address
+      ip = cloudsqlPostgresqlSinkConfig.getConnection().getConnectionName();
+    } else {
+      // connection is of the form <projectId>:<region>:<instanceName>
+      String[] connectionParams = cloudsqlPostgresqlSinkConfig.getConnection().getConnectionName().split(":");
+      ip = connectionParams[2];
+      location = connectionParams[1];
+    }
+    String fqn = DBUtils.constructFQN("postgres", ip, 0,
```
Comment (itsankit-google) [inline cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java]: similarly use `5432` for postgres

File: cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java [hunk 17]
```diff
@@ -107,6 +112,29 @@ protected FieldsValidator getFieldsValidator() {
     return new PostgresFieldsValidator();
   }
 
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSinkContext context) {
+    String ip;
```
Comment (itsankit-google) [inline cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java]: similar comment here

File: mysql-plugin/widgets/Mysql-batchsink.json [hunk 18]
```diff
@@ -87,7 +87,7 @@
           "label": "Reference Name",
           "name": "referenceName",
           "widget-attributes": {
-            "placeholder": "Name used to identify this sink for lineage"
+            "placeholder": "Name used to identify this source for lineage. Typically, the name of the table/view."
```
Comment (itsankit-google) [inline mysql-plugin/widgets/Mysql-batchsink.json]: sink?

File: oracle-plugin/widgets/Oracle-batchsink.json [hunk 19]
```diff
@@ -145,7 +145,7 @@
           "label": "Reference Name",
           "name": "referenceName",
           "widget-attributes": {
-            "placeholder": "Name used to identify this sink for lineage"
+            "placeholder": "Name used to identify this source for lineage. Typically, the name of the table/view."
```
Comment (itsankit-google) [inline oracle-plugin/widgets/Oracle-batchsink.json]: sink?

File: postgresql-plugin/widgets/Postgres-batchsink.json [hunk 20]
```diff
@@ -87,7 +87,7 @@
           "label": "Reference Name",
           "name": "referenceName",
           "widget-attributes": {
-            "placeholder": "Name used to identify this sink for lineage"
+            "placeholder": "Name used to identify this source for lineage. Typically, the name of the table/view."
```
Comment (itsankit-google) [inline postgresql-plugin/widgets/Postgres-batchsink.json]: sink?

File: cloudsql-postgresql-plugin/pom.xml [hunk 21]
```diff
@@ -134,6 +134,7 @@
           <instructions>
             <_exportcontents>
               io.cdap.plugin.cloudsql.postgres.*;
+              io.cdap.plugin.postgres.*;
```
Comment (itsankit-google) [inline cloudsql-postgresql-plugin/pom.xml:137]: why is this change needed?

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java (hunk 9 above)
Comment (snehalakshmisha) [reply cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java]: this is unused. In the UI I don't see any port in the connection?

File: cloudsql-postgresql-plugin/pom.xml (hunk 21 above)
Comment (snehalakshmisha) [reply cloudsql-postgresql-plugin/pom.xml:137]: this is because when I ran the pipeline with cloudsql postgres it gave a "PostgresDBRecord not found" exception. This is likely a bug from this PR - https://github.com/data-integrations/database-plugins/pull/282.
So adding this module in the exports fixed it.

Comment (itsankit-google) [reply cloudsql-postgresql-plugin/pom.xml:137]: Got it, thanks.

File: cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java [hunk 22]
```diff
@@ -107,6 +112,30 @@ protected FieldsValidator getFieldsValidator() {
     return new PostgresFieldsValidator();
   }
 
+  @Override
+  protected LineageRecorder getLineageRecorder(BatchSinkContext context) {
+    String host;
```
Comment (fernst) [inline cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java:117]: I noticed this logic is very similar for all CloudSQL based instances. Is it possible to consolidate this into a common method?

**Review by fernst — APPROVED** [review]
Approved, however I left one comment.

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java [hunk 23]
```diff
@@ -71,6 +76,29 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
+    if (CloudSQLUtil.PRIVATE_INSTANCE.equalsIgnoreCase(cloudsqlMysqlSinkConfig.getConnection().getInstanceType())) {
+      // connection is the private IP address
+      host = cloudsqlMysqlSinkConfig.getConnection().getConnectionName();
+    } else {
+      // connection is of the form <projectId>:<region>:<instanceName>
+      String[] connectionParams = cloudsqlMysqlSinkConfig.getConnection().getConnectionName().split(":");
+      host = connectionParams[2];
+      location = connectionParams[1];
+    }
+    String fqn = DBUtils.constructFQN("mysql", host, 3306,
```
Comment (itsankit-google) [inline cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java:93]: nit - define the constants in some `utils` or `constants` class. This comment applies to rest of the PR as well.

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java [hunk 24]
```diff
@@ -71,6 +76,29 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
+                                      cloudsqlMysqlSinkConfig.getConnection().getDatabase(),
+                                      cloudsqlMysqlSinkConfig.getReferenceName());
+    Asset.Builder assetBuilder = Asset.builder(cloudsqlMysqlSinkConfig.getReferenceName()).setFqn(fqn);
+    if (!StringUtils.isEmpty(location)) {
```
Comment (itsankit-google) [inline cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java:97]: nit - can use `Strings.isNullOrEmpty`, will eliminate the need of initialising the location as empty string. This comment applies to rest of the PR as well.

**Review by itsankit-google — APPROVED** [review]
minor nits, otherwise LGTM, thanks.

File: cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java (hunk 22 above)
Comment (itsankit-google) [reply cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java:117]: good point, can be moved to [CloudSQLUtil class](https://github.com/data-integrations/database-plugins/blob/develop/database-commons/src/main/java/io/cdap/plugin/util/CloudSQLUtil.java) if possible.

Comment (snehalakshmisha) [reply cloudsql-postgresql-plugin/src/main/java/io/cdap/plugin/cloudsql/postgres/CloudSQLPostgreSQLSink.java:117]: It needs a bit of thinking as the common method needs to accommodate multiple types of objects(the connection inside each config is different and extracting the values will mean applying the current logic here). Created PLUGIN-1420 to track this

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java (hunk 23 above)
Comment (snehalakshmisha) [reply cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java:93]: Will track as part of created PLUGIN-1420

File: cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java (hunk 24 above)
Comment (snehalakshmisha) [reply cloudsql-mysql-plugin/src/main/java/io/cdap/plugin/cloudsql/mysql/CloudSQLMySQLSink.java:97]: So if I do not initialize location the IDE complains. It doesn't happen with host because it is initialized in if and else clauses. This means in one place or the other we will have to initialize it
